File: sdk/ts/src/errcorrect.ts
/**
 * @module errcorrect
 * @package @hop-top/kit
 *
 * Factor 4 — Corrective Error Model.
 *
 * Structured errors with machine-readable code, human-readable message,
 * likely cause, suggested fix, and alternative commands.
 */

export interface CorrectedError {
  code: string;
  message: string;
  cause?: string;
  fix?: string;
  alternatives?: string[];
  retryable?: boolean;
}

/**
 * Creates an Error instance augmented with CorrectedError fields.
 * Serializes cleanly to JSON (only defined fields).
 */
export function createCorrectedError(
  opts: CorrectedError,
): Error & CorrectedError {
  const err = new Error(opts.message) as Error & CorrectedError & {
    toJSON(): Record<string, unknown>;
  };
  err.code = opts.code;
  if (opts.cause !== undefined) err.cause = opts.cause;
  if (opts.fix !== undefined) err.fix = opts.fix;
  if (opts.alternatives !== undefined) err.alternatives = opts.alternatives;
  if (opts.retryable !== undefined) err.retryable = opts.retryable;

  err.toJSON = () => {
    const o: Record<string, unknown> = {
      code: err.code,
      message: err.message,
    };
    if (err.cause !== undefined) o.cause = err.cause;
    if (err.fix !== undefined) o.fix = err.fix;
    if (err.alternatives !== undefined) o.alternatives = err.alternatives;
    if (err.retryable !== undefined) o.retryable = err.retryable;
    return o;
  };

  return err;
}

/**
 * Formats a CorrectedError for terminal display.
 *
 * ```
 *   ERROR  mission not found
 *   Cause: no mission matches "bogux"
 *   Fix:   spaced mission list
 *   Try:   spaced mission search bogux
 * ```
 */
export function formatError(
  err: CorrectedError,
  opts?: { noColor?: boolean },
): string {
  const red = opts?.noColor ? (s: string) => s : (s: string) => `\x1b[31m${s}\x1b[0m`;
  const bold = opts?.noColor ? (s: string) => s : (s: string) => `\x1b[1m${s}\x1b[0m`;

  const lines: string[] = [];
  lines.push(`  ${red(bold('ERROR'))}  ${err.message}`);

  if (err.cause) {
    lines.push(`  Cause: ${err.cause}`);
  }
  if (err.fix) {
    lines.push(`  Fix:   ${err.fix}`);
  }
  if (err.alternatives && err.alternatives.length > 0) {
    for (const alt of err.alternatives) {
      lines.push(`  Try:   ${alt}`);
    }
  }

  return lines.join('\n');
}

File: sdk/py/hop_top_kit/errcorrect.py
"""errcorrect — corrective error model (12-factor AI CLI, Factor 4).

Errors carry machine-readable code, human-readable message, likely
root cause, suggested fix, and alternatives. Supports both structured
(dict/JSON) and terminal-formatted output.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field


@dataclass
class CorrectedError(Exception):
    """Error with corrective guidance."""

    code: str
    message: str
    cause: str = ""
    fix: str = ""
    alternatives: list[str] = field(default_factory=list)
    retryable: bool = False

    def __str__(self) -> str:
        return f"{self.code}: {self.message}"

    def to_dict(self) -> dict:
        return asdict(self)

    def format_terminal(self, no_color: bool = False) -> str:
        bold = "" if no_color else "\033[1m"
        red = "" if no_color else "\033[31m"
        yellow = "" if no_color else "\033[33m"
        reset = "" if no_color else "\033[0m"

        lines = [f"{red}{bold}Error [{self.code}]{reset}: {self.message}"]

        if self.cause:
            lines.append(f"  {yellow}Cause:{reset} {self.cause}")
        if self.fix:
            lines.append(f"  {yellow}Fix:{reset} {self.fix}")
        if self.alternatives:
            lines.append(f"  {yellow}Alternatives:{reset}")
            for alt in self.alternatives:
                lines.append(f"    - {alt}")

        return "\n".join(lines)
